docs/features/interceptors.md

```diff
@@ -0,0 +1,121 @@
+    {
+    }
+}
+```
+
+#### File paths
+
+File paths used in `[InterceptsLocation]` must exactly match the paths on the syntax trees they refer to by ordinal comparison. `SyntaxTree.FilePath` has already applied `/pathmap` substitution, so the paths used in the attribute will be less environment-specific in many projects.
+
+The compiler does not map `#line` directives when determining if an `[InterceptsLocation]` attribute intercepts a particular call in syntax.
+
+#### Position
+The implementation currently uses 0-indexed line and character numbers. However, we may want to change that before shipping it as an experimental feature to be 1-indexed, to match existing places where these values are displayed to the user (e.g. `Diagnostic.ToString`).
+
+The location of the call is the location of the name syntax which denotes the interceptable method. For example, in `app.MapGet(...)`, the name syntax for `MapGet` would be considered the location of the call. If we allow intercepting calls to property accessors in the future (e.g `obj.Property`), we would also be able to use the name syntax in this way.
+
+#### Attribute creation
+
+The goal of the above decisions is to make it so that when source generators are filling in `[InterceptsLocation(...)]`, they simply need to read `nameSyntax.SyntaxTree.FilePath` and `nameSyntax.GetLineSpan().Span.Start` for the exact file path and position information they need to use.
+
+We should provide samples of recommended coding patterns for generator authors, and perhaps provide public helper methods, which will make it relatively easy for them to do the right thing. For example, we could provide a helper on InvocationExpressionSyntax which returns the values that need to be put in `[InterceptsLocation]`.
+
+### Non-invocation method usages
+
+Conversion to delegate type, address-of, etc. usages of methods cannot be intercepted.
+
+Interception can only occur for calls to ordinary member methods--not constructors, delegates, properties, local functions, etc.
+
+### Arity
+
+Interceptors cannot have type parameters or be declared in generic types at any level of nesting.
+
+### Signature matching
+
+The return and parameter types of the interceptable and interceptor methods must match exactly, except that:
+- when an interceptable instance method is compared to a classic extension method, we use the extension method in reduced form for comparison, and
+- reference types with oblivious nullability can match either annotated or unannotated reference types.
+
+Arity does not need to match between intercepted and interceptor methods. In other words, it is permitted to intercept a generic method with a non-generic interceptor.
+
+### Conflicting interceptors
+
+If more than one interceptor refers to the same location, it is a compile-time error.
+
+If an `[InterceptsLocation]` attribute is found in the compilation which does not refer to the location of an interceptable method call, it is a compile-time error.
```

Comment:
>[InterceptsLocation] [](http://example.com/codeflow?start=7&length=20)

Are the `[InterceptsLocation]` and `[Interceptable]` attributes dropped when emitting the methods to metadata?

Reply:
No. We could consider:
- having the compiler drop the attributes automatically (treating them as pseudo-custom attributes, perhaps)
- making the attribute declaration in the framework "conditional" on some debug symbol which is not usually provided.
- deliberately keeping them there. I can't think of any good reason to do this.

Perhaps the first option would be best.

Reply:
Adding a PROTOTYPE to the doc.